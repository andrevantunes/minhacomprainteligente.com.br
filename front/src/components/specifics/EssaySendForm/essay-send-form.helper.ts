export const TEN_MB = 10 * 1024 * 1024;

export const IMAGE_EXTENSIONS = ["png", "jpeg", "jpg"];

export const successRedirectPath = "/app/redacao/minhas";

export const dictionary = {
  title: "Enviar redação",
  form: {
    proposal: { title: "Tema", label: "Digite o tema aqui", name: "essay-proposal-name" },
    file: { title: "Arquivo", name: "essay-proposal-file" },
    confirmation: {
      name: "essay-proposal-confirmation",
      label:
        "Certifico que essa redação é de minha autoria e respeita os demais termos de uso do Me Salva!",
    },
    submit: {
      label: "Enviar",
      success: "Redação enviada com sucesso!",
      error: "Erro ao enviar redação, tente mais tarde.",
    },
  },
};

interface MountEssayExplanationListProps {
  essayPersonalCorrection: boolean;
  correctionStyleName?: string;
}

type MountEssayExplanationList = (props: MountEssayExplanationListProps) => string[];

export const mountEssayExplanationList: MountEssayExplanationList = ({
  essayPersonalCorrection,
  correctionStyleName,
}) => {
  const type = essayPersonalCorrection ? "Correção personalizada" : "Correção padrão";
  return [
    "A correção será feita com o prazo de <strong>10 dias úteis</strong> após o envio.",
    correctionStyleName
      ? `A correção será feita com o estilo de <strong>${correctionStyleName}</strong>`
      : "Esta redação será corrigida segundo os critérios do tema selecionado",
    `Tipo de correção: <strong>${type}</strong>`,
    "A redação será enviada junto com o rascunho, caso tenha, mas o rascunho não faz parte da correção.",
  ];
};

export const getCorrectionType = (essayPersonalCorrection = false) =>
  essayPersonalCorrection ? "redacao-personalizada" : "redacao-padrao";
