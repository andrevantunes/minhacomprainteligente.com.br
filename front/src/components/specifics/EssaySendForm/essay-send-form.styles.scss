.essay-send-form {
  &__form {
    display: flex;
    flex-direction: column;
    gap: $spacing-xl;
    &-list {
      margin-bottom: 0;
      margin-left: $spacing-md;
    }
  }
  &--is-disabled {
    color: $color-action-disabled;
    pointer-events: none;
  }
}
