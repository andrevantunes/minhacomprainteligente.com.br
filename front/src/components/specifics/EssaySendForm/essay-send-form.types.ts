interface EssaySendFormProps {}

export type { EssaySendFormProps };
