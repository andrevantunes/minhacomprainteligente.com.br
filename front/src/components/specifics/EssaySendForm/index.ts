export { default as EssaySendForm } from "./essay-send-form.component";
export * from "./essay-send-form.types";
