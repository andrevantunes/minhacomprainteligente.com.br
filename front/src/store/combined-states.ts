import { defaultState as accesses } from "./AccessesStore/accesses-store.state";
import { defaultState as user } from "./UserStore/user-store.state";

// plop import

const combinedStates = {
  ...user,
  ...accesses,
  // plop export
};

export default combinedStates;
