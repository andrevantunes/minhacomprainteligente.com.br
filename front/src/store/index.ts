export * from "./store.global";
export * from "./store.helper";
export * from "./store.types";

export * as AccessesStore from "./AccessesStore/accesses-store.action";
export * as UserStore from "./UserStore/user-store.action";
export * as CartStore from "./CartStore/cart-store.action";
