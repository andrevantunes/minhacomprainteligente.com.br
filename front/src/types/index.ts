export * from "./accesses.types";
export * from "./cart.types";
export * from "./auth.types";
export * from "./component.types";
export * from "./events.types";
export * from "./page.types";
export * from "./platform.types";
export * from "./templates.types";
export * from "./user.types";
