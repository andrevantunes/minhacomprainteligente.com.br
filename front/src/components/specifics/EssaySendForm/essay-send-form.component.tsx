import { Title } from "@/components/adapters/Title";
import { Text } from "@/components/adapters/Text";
import { DangerousHTML } from "@/components/basics/DangerousHTML";
import { findEssayProposals, getEssayProposals, sendEssayProposal } from "@/requests";
import { useStore, StoreType } from "@/store";
import {
  Card,
  Checkbox,
  FileField,
  List,
  Notification,
  SelectField,
  SelectFieldOption,
  Skeleton,
  SkeletonVariants,
  SubmitButton,
  TextField,
} from "@andrevantunes/andrevds";
import classNames from "classnames";
import React, { useEffect, useState } from "react";
import useSWR from "swr";
import {
  dictionary,
  getCorrectionType,
  IMAGE_EXTENSIONS,
  mountEssayExplanationList,
  successRedirectPath,
  TEN_MB,
} from "./essay-send-form.helper";
import { Button } from "@/components/adapters/Button";
import { useRouter } from "next/router";
import { removeBasePath } from "@/helpers/links.helpers";
import * as AccessStore from "@/store/AccessesStore/accesses-store.action";

const EssaySendForm = () => {
  const [{ essayPersonalCorrection, essayCredits, essayUnlimited }] = useStore(StoreType.Accesses);
  const hasEssay = essayCredits > 0 || essayUnlimited;

  const router = useRouter();
  const { token: mediumToken, battery: batteryToken } = router?.query || {};

  const { data: options, error } = useSWR("essays/proposals", getEssayProposals);

  const [file, setFile] = useState<string | null>(null);
  const [proposal, setProposal] = useState<SelectFieldOption | undefined>(undefined);
  const [confirmation, setConfirmation] = useState(false);
  const [submitting, setSubmitting] = useState(false);
  const [singleProposal, setSingleProposal] = useState<any>(null);

  const cn = classNames("essay-send-form", { "essay-send-form--is-disabled": !hasEssay });
  const isLoading = !options && !singleProposal;
  const correctionStyleName = proposal?.value?.correctionStyleName;
  const correctionType = getCorrectionType(essayPersonalCorrection);

  const handleSubmit = async (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    setSubmitting(true);
    try {
      await sendEssayProposal({
        file,
        correctionType,
        mediumToken,
        batteryToken,
        ...proposal?.value,
      });
      AccessStore.decreaseEssayCredits();
      Notification.success(dictionary.form.submit.success);
      router.push(removeBasePath(successRedirectPath));
    } catch (error: any) {
      Notification.error(error?.message || dictionary.form.submit.error);
      setSubmitting(false);
    }
  };

  useEffect(() => {
    if (mediumToken) {
      findEssayProposals(String(mediumToken)).then(setSingleProposal);
    }
  }, [mediumToken]);

  if (error) {
    return (
      <Card className={cn} elevation="md">
        <Title size="xs">Não foi possível carregar as propostas de redação :(</Title>
        <Text className="my-md">Tente mais tarde ou entre em contato com o nosso time.</Text>
        <Button variant="neutral" size="sm" onClick={location.reload}>
          Tentar novamente
        </Button>
      </Card>
    );
  }

  return (
    <Card className={cn} elevation="md">
      <form data-testid="essay-send-form" className="essay-send-form__form" onSubmit={handleSubmit}>
        <input type="hidden" value={batteryToken} name="battery_token" />
        <div>
          <Skeleton active={isLoading} width={200}>
            <Title level={3} size="xs" className="mb-sm">
              {dictionary.form.proposal.title}
            </Title>
          </Skeleton>
          <Skeleton active={isLoading}>
            {!singleProposal && (
              <SelectField
                defaultOption={proposal}
                disabled={!hasEssay || submitting}
                autoComplete="off"
                name={dictionary.form.proposal.name}
                required
                label={dictionary.form.proposal.label}
                enableFilter
                options={options || []}
                onSelect={setProposal}
              />
            )}
            {singleProposal && (
              <>
                <input
                  type="hidden"
                  value={singleProposal.id}
                  name={dictionary.form.proposal.name}
                />
                <TextField
                  disabled={true}
                  autoComplete="off"
                  value={singleProposal.name}
                  label="Proposta selecionada"
                />
              </>
            )}
          </Skeleton>
        </div>
        <div>
          <Skeleton active={isLoading} width={200}>
            <Title level={3} size="xs" className="mb-sm">
              {dictionary.form.file.title}
            </Title>
          </Skeleton>
          <Skeleton active={isLoading}>
            <FileField
              data-testid={dictionary.form.file.name}
              name={dictionary.form.file.name}
              disabled={!hasEssay || submitting}
              maxSize={TEN_MB}
              onUploadFile={setFile}
              extensions={IMAGE_EXTENSIONS}
            />
          </Skeleton>
        </div>
        <Skeleton active={isLoading} variant={SkeletonVariants.Paragraph}>
          <List
            className="essay-send-form__form-list"
            defaultBulletIconName="checkmark"
            list={mountList({ correctionStyleName, essayPersonalCorrection })}
          />
        </Skeleton>
        <div>
          <Skeleton active={isLoading}>
            <Checkbox
              disabled={!hasEssay || submitting}
              defaultChecked={confirmation}
              onClick={() => setConfirmation(!confirmation)}
              name={dictionary.form.confirmation.name}
              label={dictionary.form.confirmation.label}
            />
          </Skeleton>
        </div>
        <Skeleton active={isLoading}>
          <SubmitButton
            label={dictionary.form.submit.label}
            disabled={!hasEssay || (!proposal && !singleProposal) || !file || !confirmation}
            iconName={submitting ? undefined : "send"}
            submitting={submitting}
          />
        </Skeleton>
      </form>
    </Card>
  );
};

const mountList = (props: { essayPersonalCorrection: boolean; correctionStyleName?: string }) => {
  const list = mountEssayExplanationList(props);
  return list.map((item) => ({
    children: <DangerousHTML>{item}</DangerousHTML>,
  }));
};

export default EssaySendForm;
