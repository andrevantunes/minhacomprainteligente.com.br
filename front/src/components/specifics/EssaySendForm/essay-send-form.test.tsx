import type { EssaySendFormProps } from "./essay-send-form.types";

import { render } from "@testing-library/react";

import EssaySendForm from "./essay-send-form.component";
import * as AccessStore from "@/store/AccessesStore/accesses-store.action";
import { ActionName } from "@/store/store.types";

const essayCreditsMock = 1;
AccessStore.update({ essayCredits: essayCreditsMock }, ActionName.Update);

jest.mock("@/requests", () => {
  return {
    getEssayProposals: async () => [{ label: "any_proposal_label", value: "any_proposal_value" }],
    sendEssayProposal: async () => jest.fn().mockReturnValue(""),
  };
});

const makeSut = (props?: EssaySendFormProps) => render(<EssaySendForm {...props} />);

describe("EssaySendForm", () => {
  it("should render component", () => {
    const wrapper = makeSut();
    expect(wrapper.container).toMatchSnapshot();
  });
});
