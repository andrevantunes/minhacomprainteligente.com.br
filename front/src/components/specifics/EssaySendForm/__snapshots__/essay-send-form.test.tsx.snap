// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EssaySendForm should render component 1`] = `
<div>
  <section
    class="card card--elevation-lw essay-send-form section-card"
  >
    <form
      class="essay-send-form__form"
      data-testid="essay-send-form"
    >
      <input
        name="battery_token"
        type="hidden"
        value=""
      />
      <div>
        <div
          class="skeleton skeleton--is-default"
          data-testid="skeleton"
          style="width: 200px;"
        >
          <div
            class="skeleton-container"
          >
            <h3
              class="heading heading--size-xs title mb-sm"
            >
              Tema
            </h3>
          </div>
        </div>
        <div
          class="skeleton skeleton--is-default"
          data-testid="skeleton"
        >
          <div
            class="skeleton-container"
          >
            <div
              class="select-field select-field--has-filter"
            >
              <div>
                <div
                  class="field field--has-right-icon"
                >
                  <fieldset
                    class="field__fieldset"
                  >
                    <div
                      class="field__content"
                    >
                      <input
                        autocomplete="off"
                        class="field__input"
                        id="essay-proposal-name"
                        name="essay-proposal-name"
                        required=""
                        value=""
                      />
                      <span
                        class="icon select-field__toggle-icon field__icon--has-action"
                        data-name="chevron-down"
                        data-testid="chevron-down"
                        tabindex="0"
                      >
                        <svg
                          fill="none"
                          height="24"
                          viewBox="0 0 24 24"
                          width="24"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M15.4676 9.76651L11.9955 13.3018L8.52349 9.76651C8.1745 9.41116 7.61074 9.41116 7.26174 9.76651C6.91275 10.1219 6.91275 10.6959 7.26174 11.0513L11.3691 15.2335C11.7181 15.5888 12.2819 15.5888 12.6309 15.2335L16.7383 11.0513C17.0872 10.6959 17.0872 10.1219 16.7383 9.76651C16.3893 9.42027 15.8166 9.41116 15.4676 9.76651Z"
                            fill="currentColor"
                          />
                        </svg>
                      </span>
                    </div>
                    <label
                      class="field__label"
                      for="essay-proposal-name"
                    >
                      Digite o tema aqui
                    </label>
                    <legend
                      class="field__legend"
                    >
                      Digite o tema aqui
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="skeleton skeleton--is-default"
          data-testid="skeleton"
          style="width: 200px;"
        >
          <div
            class="skeleton-container"
          >
            <h3
              class="heading heading--size-xs title mb-sm"
            >
              Arquivo
            </h3>
          </div>
        </div>
        <div
          class="skeleton skeleton--is-default"
          data-testid="skeleton"
        >
          <div
            class="skeleton-container"
          >
            <div
              class="file-field"
            >
              <div
                class="file-field__content text-center"
              >
                <img
                  class="file-field__content__image file-field__content__image--is-default"
                  height="164"
                  src="https://cdn.mesalva.com/uploads/image/MjAyMi0wNS0xNiAxOTozNDozMiArMDAwMDQ5NTY5MQ%3D%3D%0A.png"
                />
                <h5
                  class="heading heading--size-xs"
                >
                  Toque para escolher um arquivo
                </h5>
              </div>
              <input
                class="file-input"
                data-testid="essay-proposal-file"
                name="essay-proposal-file"
                type="file"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        data-testid="skeleton"
      >
        <div
          class="skeleton skeleton--is-paragraph"
          data-testid="skeleton-paragraph"
          style="width: 100%;"
        />
        <div
          class="skeleton skeleton--is-paragraph"
          data-testid="skeleton-paragraph"
          style="width: 100%;"
        />
        <div
          class="skeleton skeleton--is-paragraph"
          data-testid="skeleton-paragraph"
          style="width: 90%;"
        />
        <div
          class="skeleton skeleton--is-paragraph"
          data-testid="skeleton-paragraph"
          style="width: 30%;"
        />
      </div>
      <div>
        <div
          class="skeleton skeleton--is-default"
          data-testid="skeleton"
        >
          <div
            class="skeleton-container"
          >
            <div
              class="checkbox"
              data-testid="checkbox"
              name="essay-proposal-confirmation"
            >
              <label
                class="checkbox__container"
                data-testid="checkbox-label"
              >
                <div
                  class="checkbox__icon"
                >
                  <span
                    class="icon"
                    data-name="square"
                    data-testid="square"
                  >
                    <svg
                      fill="none"
                      height="24"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        clip-rule="evenodd"
                        d="M6 21H18C19.6569 21 21 19.6569 21 18V6C21 4.34315 19.6569 3 18 3H6C4.34315 3 3 4.34315 3 6V18C3 19.6569 4.34315 21 6 21ZM5 6C5 5.44772 5.44772 5 6 5H18C18.5523 5 19 5.44772 19 6V18C19 18.5523 18.5523 19 18 19H6C5.44772 19 5 18.5523 5 18V6Z"
                        fill="currentColor"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </span>
                </div>
                <input
                  class="checkbox__input"
                  data-testid="checkbox-input"
                  id="essay-proposal-confirmation"
                  type="checkbox"
                />
                <span
                  class="text text--size-sm checkbox__label"
                >
                  Certifico que essa redação é de minha autoria e respeita os demais termos de uso do Me Salva!
                </span>
              </label>
            </div>
          </div>
        </div>
      </div>
      <div
        class="skeleton skeleton--is-default"
        data-testid="skeleton"
      >
        <div
          class="skeleton-container"
        >
          <button
            class="btn btn--primary btn--size-md btn--is-disabled submit-button"
            disabled=""
            type="submit"
          >
            <span
              class="icon btn__icon"
              data-name="send"
              data-testid="send"
            >
              <svg
                height="24"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M3.4 20.4L20.85 12.92C21.66 12.57 21.66 11.43 20.85 11.08L3.4 3.60002C2.74 3.31002 2.01 3.80002 2.01 4.51002L2 9.12002C2 9.62002 2.37 10.05 2.87 10.11L17 12L2.87 13.88C2.37 13.95 2 14.38 2 14.88L2.01 19.49C2.01 20.2 2.74 20.69 3.4 20.4Z"
                  fill="currentColor"
                />
              </svg>
            </span>
            <span
              class="btn__content"
            >
              Enviar
            </span>
          </button>
        </div>
      </div>
    </form>
  </section>
</div>
`;
